### PR TITLE
ui: mutable deployments

### LINKED ISSUE #1239: Mutable Deployment Support in the browser UI
**Is your feature request related to a problem? Please describe.**
Alongside the core Mutable Deployment Support feature (https://github.com/hashicorp/waypoint/issues/1235), we need to support that functionality in the UI

**Describe the solution you'd like**
- surface the mutability of deployments in the UI
- adapt the UI to decouple per-deploy horizon URLs from releases
- surface "generation" objects for operations

### PR DESCRIPTION
**Pull request by jgwhite**
## What is this?

Addresses #1239

Adds the UI to display _mutable deployments_.  

_Mutable deployments_ are groups of deployment versions that utilize the same resources, where newer deployments in the same _generation_ overwrite previous deployments (ie only latest version is deployed at the same deployment URL). This functionality was added to accommodate users' existing workflows.

## What’s the plan?

- [x] Create good fixture data in Mirage
- [x] Implement the designs
- [x] Ensure immutable deployments still render well
- [x] Ensure destroyed deployments still render well
- [x] Use the `{{t ...}}` helper for all text
- [x] Write some light tests (we can probably borrow the dev Mirage config)
- [x] Changelog entry

## What does it look like?

Screen Shot 2021-06-02 at 1 17 39 PM  

> Deployments list without destroyed deployments, note the grayed background of overwritten deployments + link to deployment only on latest

Screen Shot 2021-06-02 at 1 17 47 PM  

> Deployments list with destroyed deployment

## How do I verify it?

1. Check out this branch
2. `yarn ember serve`
3. Visit `http://localhost:4200/tests`
4. Verify all tests pass
5. Visit `http://localhost:4200`
6. Check the deployments under `mutable-application` in `mutable-project`
7. Verify there's no buggy behavior and everything else is untouched (check the builds section)

### CODE REVIEW

File: ui/app/templates/workspace/projects/project/app/deployments.hbs [hunk 1]
```diff
@@ -15,20 +19,20 @@
...
+      <Pds::Button
+        data-test-display-destroyed-button
+        @compact={{true}}
+        @variant="ghost"
+        {{on "click" this.showDestroyed}}>
+        Display destroyed deployments
+      </Pds::Button>
+    </Pds::HelpText>
+  </div>
+{{/if}}
```
Comment (jgwhite) [inline ui/app/templates/workspace/projects/project/app/deployments.hbs:36]: Very glad to see this move outside of the `<ul>`.

**Comment by jgwhite** [comment]
Technically I can’t approve this PR because I opened it, but it gets my 👍 nonetheless.

**Review by briancain — APPROVED** [review]
Looks good to me 👍🏻 I was able to build my own waypoint server container with the latest changes to test. Ran through a nomad jobspec and a k8s apply project and did a few waypoint ups for both.

File: ui/app/templates/workspace/projects/project/app/deployments.hbs [hunk 2]
```diff
@@ -2,9 +2,13 @@
...
+<ul data-test-deployment-list class="list">
+  {{#each this.deploymentsByGeneration key="generationID" as |group|}}
+    {{#each group.deployments key="id" as |deployment|}}
+      {{#if (or this.isShowingDestroyed (not-eq deployment.state 4))}}
+        <AppItem::Deployment @deployment={{deployment}} @latest={{group.deployments.[0]}} />
+      {{/if}}
+    {{/each}}
   {{else}}
     <EmptyState>
       <p>There are no deployments to display for this app yet</p>
```
Comment (almonk) [inline ui/app/templates/workspace/projects/project/app/deployments.hbs:14]: Should this text be in the `translations/` file?

File: ui/app/styles/_app-item.scss [hunk 3]
```diff
@@ -64,4 +65,18 @@
...
+      text-transform: inherit;
+      margin-left: scale.$sm-2;
+    }
+  }
+
+  &--previous {
+    background: color.$ui-cool-gray-050;
+    background-clip: border-box;
+    margin-bottom: 0px;
+    border-top: 1rem solid color.$ui-cool-gray-050 !important;
```
Comment (almonk) [inline ui/app/styles/_app-item.scss:80]: Ideally we wouldn't use `!important`

File: ui/app/templates/workspace/projects/project/app/deployments.hbs [hunk 4]
```diff
@@ -15,20 +19,20 @@
...
+{{#if (and this.hasMoreDeployments (not this.isShowingDestroyed))}}
+  <div class="destroyed-deployments-filter">
+    <Pds::HelpText>
+      {{t "page.deployments.destroyed_and_hidden_state"}}
+      <Pds::Button
+        data-test-display-destroyed-button
+        @compact={{true}}
+        @variant="ghost"
+        {{on "click" this.showDestroyed}}>
+        Display destroyed deployments
```
Comment (almonk) [inline ui/app/templates/workspace/projects/project/app/deployments.hbs:32]: I'm not sure if we're attempting to move all strings to `{{t ...` but if so there's another case here

File: ui/translations/en-us.yaml [hunk 5]
```diff
@@ -21,6 +21,16 @@ page:
     title: 'Builds'
   deployments:
     title: 'Deployments'
+    destroyed_label: 'Destroyed'
+    external_url_label: 'Visit Deployment'
+    replaced_label: 'Replaced by '
+    empty_state: 'There are no deployments to display for this app yet'
+    destroyed_and_hidden_state: 'Some deployments have been destroyed and are hidden.'
```
Comment (almonk) [inline ui/translations/en-us.yaml:28]: I'd say remove the trailing period from this sentence as we typically don't end single sentences with a period in the UI

**Review by almonk — CHANGES_REQUESTED** [review]
Looks great and the attention to detail is :chefskiss:

Some minor changes highlighted that I think we should make

File: ui/app/templates/workspace/projects/project/app/deployments.hbs (hunk 2 above)
Comment (sabrinako) [reply ui/app/templates/workspace/projects/project/app/deployments.hbs:14]: I started moving some of the text in this `deployments.hbs` file to the translations files, but @jgwhite said the checklist item about it on this PR was meant for the single deployment item template, not this overarching deployments template.

Comment (sabrinako) [reply ui/app/templates/workspace/projects/project/app/deployments.hbs:14]: I added the other text to the translation file. Jamie and I discussed this specific block of text yesterday and agreed the paragraph breaks make it a bit more of a complex situation. It's a little out of scope for this PR, so we can add this text to the localization texts some other time.

**Review by almonk — APPROVED** [review]

**Pull request merged**